src/main/java/net/neoforged/neoforge/client/model/data/ModelData.java

```diff
@@ -63,25 +71,40 @@ public static Builder builder() {
     }
 
     public static final class Builder {
-        private final Map<ModelProperty<?>, Object> properties = new IdentityHashMap<>();
+        /**
+         * Hash maps are slower than array maps for *extremely* small maps (empty maps or singletons are the most
+         * extreme examples). Many block entities/models only use a single model data property, which means the
+         * overhead of hashing is quite wasteful. However, we do want to support any number of properties with
+         * reasonable performance. Therefore, we use an array map until the number of properties exceeds this
+         * threshold, at which point we convert it to a hash map.
+         */
+        private static final int HASH_THRESHOLD = 4;
+
+        private Map<ModelProperty<?>, Object> properties;
 
         private Builder(@Nullable ModelData parent) {
             if (parent != null) {
-                properties.putAll(parent.properties);
+                // When cloning the map, use the expected type based on size
+                properties = parent.properties.size() >= HASH_THRESHOLD ? new Reference2ReferenceOpenHashMap<>(parent.properties) : new Reference2ReferenceArrayMap<>(parent.properties);
+            } else {
+                properties = new Reference2ReferenceArrayMap<>();
             }
         }
 
         @Contract("_, _ -> this")
         public <T> Builder with(ModelProperty<T> property, T value) {
             Preconditions.checkState(property.test(value), "The provided value is invalid for this property.");
             properties.put(property, value);
+            // Convert to a hash map if needed
+            if (properties.size() >= HASH_THRESHOLD && properties instanceof Reference2ReferenceArrayMap<ModelProperty<?>, Object>) {
```

Comment:
I feel like this is an unnecessary amount of instance of operations on the builder, which could make this change more expensive in cases where model data is changing often. Why not just change the map type in the build method?

Reply:
XFact noted that the `>=` can safely be changed to `==`, which should eliminate the `instanceof` most of the time for larger objects.